// crossbow/dist/command.tasks.js
"use strict";
var command_run_1 = require("./command.run");
var task_resolve_1 = require("./task.resolve");
var Rx = require("rx");
var file_utils_1 = require("./file.utils");
var task_utils_1 = require("./task.utils");
function execute(trigger) {
    var input = trigger.input, config = trigger.config;
    var allNames = Object.keys(trigger.input.tasks);
    var possibleParents = allNames.filter(function (x) { return task_utils_1.isParentGroupName(x); });
    var possibleDefaults = allNames.filter(function (x) { return !task_utils_1.isParentGroupName(x); });
    var cliInput = trigger.cli.input.slice(1);
    var defaultInputNames = cliInput.filter(function (x) { return !task_utils_1.isParentRef(x, possibleParents); });
    var parentInputNames = cliInput.filter(function (x) { return task_utils_1.isParentRef(x, possibleParents); });
    /**
     * Either resolve ALL tasks, or a subset if given
     * via the cli.
     *
     * eg:
     *      crossbow ls -> all tasks
     *      crossbow ls build-all -> only build all tasks
     */
    var defaultsToResolve = (function () {
        if (defaultInputNames.length) {
            return defaultInputNames;
        }
        if (parentInputNames.length) {
            return [];
        }
        var taskNamesFromTasksDir = file_utils_1.getPossibleTasksFromDirectories(config.tasksDir, config.cwd);
        return possibleDefaults.concat(taskNamesFromTasksDir);
    })();
    var parentsToResolve = (function () {
        if (parentInputNames.length) {
            return parentInputNames;
        }
        if (defaultInputNames.length) {
            return [];
        }
        return possibleParents;
    })();
    /**
     * Resolve the subset
     * @type {Tasks}
     */
    var resolvedDefault = task_resolve_1.resolveTasks(defaultsToResolve, trigger);
    var resolvedParents = parentsToResolve
        .map(task_utils_1.getChildName)
        .reduce(function (acc, key) {
        var childKeys = Object.keys(task_utils_1.getChildItems(key, input.tasks));
        var plainName = key.slice(1, -1);
        if (!acc[plainName]) {
            acc[plainName] = childKeys;
        }
        else {
            acc[plainName].push.apply(acc[plainName], childKeys);
        }
        return acc;
    }, {});
    function getParents(resolvedParents, trigger) {
        return Object.keys(resolvedParents).map(function (key) {
            var parent = resolvedParents[key];
            var items = parent.map(function (x) { return key + ":" + x; });
            var resolved = task_resolve_1.resolveTasks(items, trigger);
            return { title: key, tasks: resolved };
        });
    }
    var groups = (function () {
        if (resolvedDefault.all.length) {
            return [
                { title: "Default Tasks", tasks: resolvedDefault }
            ].concat(getParents(resolvedParents, trigger));
        }
        return getParents(resolvedParents, trigger);
    })();
    var tasks = groups.reduce(function (acc, group) {
        return acc.concat(group.tasks.all);
    }, []);
    return Rx.Observable.just({
        setup: {
            groups: groups,
            tasks: tasks,
            errors: []
        }
    });
}
function handleIncomingTasksCommand(cli, input, config, reporter) {
    return execute({
        cli: cli,
        input: input,
        config: config,
        reporter: reporter,
        type: command_run_1.TriggerTypes.command
    });
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = handleIncomingTasksCommand;
