// crossbow/dist/index.js
#!/usr/bin/env node
"use strict";
var config_1 = require("./config");
var file_utils_1 = require("./file.utils");
var input_resolve_1 = require("./input.resolve");
var reports = require("./reporter.resolve");
var Rx = require("rx");
var setup_envFile_1 = require("./setup.envFile");
var setup_bin_1 = require("./setup.bin");
var fs = require('fs');
var _ = require("../lodash.custom");
var debug = require("debug")("cb:init");
var availableCommands = {
    run: "./command.run",
    r: "./command.run",
    tasks: "./command.tasks",
    t: "./command.tasks",
    ls: "./command.tasks",
    watch: "./command.watch",
    w: "./command.watch",
    watchers: "./command.watchers",
    init: "./command.init",
    docs: "./command.docs",
};
var isCommand = function (input) { return Object.keys(availableCommands).indexOf(input) > -1; };
/**
 * This the the proxy that allows command/run mode to be handled
 * @param preparedInput
 */
function handleIncoming(preparedInput) {
    var cli = preparedInput.cli, userInput = preparedInput.userInput, config = preparedInput.config, reportFn = preparedInput.reportFn;
    // if the user provided a --cbfile flag, the type 'CBFile'
    // must be available, otherwise this is an error state
    if (userInput.type === input_resolve_1.InputTypes.CBFile) {
        return handleCBfileMode(cli, config, reportFn);
    }
    return processInput(cli, userInput.inputs[0], config, reportFn);
}
exports.handleIncoming = handleIncoming;
function handleCBfileMode(cli, config, reportFn) {
    var createFilePaths = file_utils_1.getRequirePaths(config);
    var input = require(createFilePaths.valid[0].resolved);
    input.default.config = processConfigs(_.merge({}, config, input.default.config), cli.flags);
    input.default.cli = cli;
    input.default.reporter = reportFn;
    if (isCommand(cli.input[0])) {
        return require(availableCommands[cli.input[0]]).default.call(null, cli, input.default, input.default.config, reportFn);
    }
    cli.input = ["run"].concat(cli.input);
    return require(availableCommands["run"]).default.call(null, cli, input.default, input.default.config, reportFn);
}
/**
 * Now decide who should handle the current command
 */
function processInput(cli, input, config, reportFn) {
    var firstArg = cli.input[0];
    return require(availableCommands[firstArg]).default.call(null, cli, input, config, reportFn);
}
function processConfigs(config, flags) {
    var cbConfig = _.merge({}, config, flags);
    return config_1.merge(cbConfig);
}
/**
 * This is the default export that can be
 * used as a convenience method.
 * Note: types are lost when using this method.
 */
var mergeConfigs = function (userInput, merged, flags) {
    return userInput.type !== input_resolve_1.InputTypes.StubInlineObject && userInput.type !== input_resolve_1.InputTypes.CBFile
        ? file_utils_1.Right(config_1.merge(_.merge({}, userInput.inputs[0].config, flags)))
        : file_utils_1.Right(merged);
};
var getConfig = function (flags, input) {
    return file_utils_1.Right(config_1.merge(flags))
        .chain(function (merged) { return getUserInput(merged, input)
        .chain(function (userInput) {
        return mergeConfigs(userInput, merged, flags).map(function (config) { return ({ config: config, userInput: userInput }); });
    }); });
};
var getUserInput = function (merged, input) {
    return file_utils_1.Right(input_resolve_1.getInputs(merged, input))
        .chain(function (userInput) { return userInput.errors.length
        ? file_utils_1.Left({ type: reports.ReportTypes.InputError, data: userInput })
        : file_utils_1.Right(userInput); });
};
var addReporters = function (config) {
    return file_utils_1.Right(reports.getReporters(config))
        .chain(function (reporters) {
        return reporters.invalid.length
            ? file_utils_1.Left({ type: reports.ReportTypes.InvalidReporter, data: { reporters: reporters } })
            : file_utils_1.Right(reporters.valid);
    });
};
var getReportFn = function (reporters) { return function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return reporters.forEach(function (x) { return x.callable.apply(null, args); });
}; };
/**
 * Handle any type of init. It could be from the CLI, or via the API.
 * eg, any command from the CLI ultimately ends up in the following call
 *    $  crossbow run task1 -c conf/cb.js
 *    -> handleIncoming({
 *          input: ['run', 'task1'],
 *          flags: {c: 'conf/cb.js'}
 *       });
 */
function getSetup(cli, input) {
    return getConfig(cli.flags, input)
        .chain(function (setup) {
        return setup_bin_1.addBinLookupsToObject(setup.config)
            .chain(function (config) { return setup_envFile_1.addEnvFilesToObject(config); })
            .chain(function (config) { return addReporters(config)
            .map(function (reporters) {
            return {
                reporters: reporters,
                config: config,
                userInput: setup.userInput,
                cli: cli,
                reportFn: getReportFn(reporters)
            };
        }); });
    });
}
exports.getSetup = getSetup;
function default_1(cli, input) {
    return getSetup(cli, input)
        .fold(function (e) {
        return Rx.Observable.just({
            errors: [e]
        });
    }, function (prepared) {
        return handleIncoming(prepared);
    });
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
